polkadot/node/core/pvf/prepare-worker/src/lib.rs

```diff
@@ -324,3 +206,157 @@ fn runtime_construction_check(
 		.map(|_runtime| ())
 		.map_err(|err| PrepareError::RuntimeConstruction(format!("{:?}", err)))
 }
+
+#[derive(Encode, Decode)]
+struct Response {
+	artifact: CompiledArtifact,
+	memory_stats: MemoryStats,
+}
+
+async fn handle_child_process(
+	pvf: PvfPrepData,
+	mut pipe_write: os_pipe::PipeWriter,
+	preparation_timeout: Duration,
+	prepare_job_kind: PrepareJobKind,
+	executor_params: ExecutorParams,
+) -> ! {
+	nix::sys::resource::setrlimit(
+		Resource::RLIMIT_CPU,
+		preparation_timeout.as_secs(),
+		preparation_timeout.as_secs(),
+	)
```

Comment:
Right now if we hit this limit, the process will be killed. How do we handle this? Does the `terminates_on_timeout` test still pass? 

I think we'll need to add back the `getrlimit` logic that you had and then removed. Otherwise, we have no way of knowing if the process was killed due to exceeding the limit or not. It will be needed for "1. Timeout due to exceeded CPU time" in https://github.com/paritytech/polkadot-sdk/issues/660.

Reply:
`setrlimit` send a `SIGXCPU` signal when the soft limit is reached and `SIGKILL` when the hard limit is reached, I'm using the `SIGXCPU` signal to handle the timeout, the same value is being used for soft and hard limit I decided to use the `SIGXCPU because it is more specific. Currently `terminates_on_timeout` is passing but I'll double check this.

Reply:
Based on the [manpage](https://linux.die.net/man/2/setrlimit) I think reaching the hard limit should kill the process?

> CPU time limit in seconds. When the process reaches the soft limit, it is sent a SIGXCPU signal. The default action for this signal is to terminate the process. However, the signal can be caught, and the handler can return control to the main program. If the process continues to consume CPU time, it will be sent SIGXCPU once per second until the hard limit is reached, at which time it is sent SIGKILL. (This latter point describes Linux behavior. Implementations vary in how they treat processes which continue to consume CPU time after reaching the soft limit. Portable applications that need to catch this signal should perform an orderly termination upon first receipt of SIGXCPU.) 

This could of course be confirmed with testing. But I think it would be cleaner to use `getrlimit` with `SELF_CHILDREN` to detect if we surpassed the limit, as we were doing previously. Signals are kind of an ugly feature and have lots of edge cases. What do you think?

Reply:
I agree, I'm going to add back the `getrlimit`.